src/cdk/datetime/datepicker.spec.ts

```diff
@@ -0,0 +1,646 @@
+    it('should throw when created', () => {
+      expect(() => createComponent(StandardDatepicker))
+          .toThrowError(/CdkDatepicker: No provider found for .*/);
+    });
+  });
+});
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d" [value]="date">
+    <cdk-datepicker #d [disabled]="disabled"></cdk-datepicker>
+  `,
+})
+class StandardDatepicker {
+  disabled = false;
+  date: Date | null = new Date(2020, JAN, 1);
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+  @ViewChild(CdkDatepickerInput) datepickerInput: CdkDatepickerInput<Date>;
+}
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d"><input [cdkDatepicker]="d"><cdk-datepicker #d></cdk-datepicker>
+  `,
+})
+class MultiInputDatepicker {}
+
+
+@Component({
+  template: `<cdk-datepicker #d></cdk-datepicker>`,
+})
+class NoInputDatepicker {
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+}
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d" [value]="date">
+    <cdk-datepicker #d [startAt]="startDate"></cdk-datepicker>
+  `,
+})
+class DatepickerWithStartAt {
+  date = new Date(2020, JAN, 1);
+  startDate = new Date(2010, JAN, 1);
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+}
+
+
+@Component({
+  template: `
+    <input [(ngModel)]="selected" [cdkDatepicker]="d">
+    <cdk-datepicker #d></cdk-datepicker>
+  `,
+})
+class DatepickerWithNgModel {
+  selected: Date | null = null;
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+  @ViewChild(CdkDatepickerInput) datepickerInput: CdkDatepickerInput<Date>;
+}
+
+
+@Component({
+  template: `
+    <input [formControl]="formControl" [cdkDatepicker]="d">
+    <cdk-datepicker #d></cdk-datepicker>
+  `,
+})
+class DatepickerWithFormControl {
+  formControl = new FormControl();
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+  @ViewChild(CdkDatepickerInput) datepickerInput: CdkDatepickerInput<Date>;
+}
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d" [(ngModel)]="date" [min]="minDate" [max]="maxDate">
+    <cdk-datepicker #d></cdk-datepicker>
+  `,
+})
+class DatepickerWithMinAndMaxValidation {
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+  date: Date | null;
+  minDate = new Date(2010, JAN, 1);
+  maxDate = new Date(2020, JAN, 1);
+}
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d" [(ngModel)]="date" [cdkDatepickerFilter]="filter">
+    <cdk-datepicker #d></cdk-datepicker>
+  `,
+})
+class DatepickerWithFilterAndValidation {
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+  date: Date;
+  filter = (date: Date) => date.getDate() != 1;
+}
+
+
+@Component({
+  template: `
+    <input [cdkDatepicker]="d" (change)="onChange()" (input)="onInput()"
+        (dateChange)="onDateChange()" (dateInput)="onDateInput()">
+    <cdk-datepicker #d></cdk-datepicker>
+  `
+})
+class DatepickerWithChangeAndInputEvents {
+  @ViewChild('d') datepicker: CdkDatepicker<Date>;
+
+  onChange() {}
```

Comment:
Rather than creating these as noops and then calling `spyOn` in the tests, you can make them as spies to begin with:

```
onChange = jasmine.createSpy('onChange spy');
onInput = jasmine.createSpy('onInput spy');

...
```